File: Puppet/Puppet/MSCustomPropertiesViewController.m
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 */

#import "AppCenter.h"
#import "MSAppCenterPrivate.h"
#import "MSCustomPropertiesViewController.h"
#import "MSCustomPropertyTableViewCell.h"

static NSInteger kPropertiesSection = 0;

@interface MSCustomPropertiesViewController ()

@property (nonatomic) NSInteger propertiesCount;

@end

@implementation MSCustomPropertiesViewController

#pragma mark - view controller

- (void)viewDidLoad {
  [super viewDidLoad];
  [self.tableView setEditing:YES animated:NO];
}

- (IBAction)send {
  MSCustomProperties *customProperties = [MSCustomProperties new];
  for (int i = 0; i < self.propertiesCount; i++) {
    MSCustomPropertyTableViewCell *cell = [self.tableView cellForRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:kPropertiesSection]];
    [cell setPropertyTo:customProperties];
  }
  [MSAppCenter setCustomProperties:customProperties];
  
  // Clear the list
  self.propertiesCount = 0;
  [self.tableView reloadData];
}

- (NSString *)cellIdentifierForRowAtIndexPath:(NSIndexPath *)indexPath {
  NSString *cellIdentifier = nil;
  if (![self isPropertiesRowSection:indexPath.section]) {
    cellIdentifier = @"send";
  } else if ([self isInsertRowAtIndexPath:indexPath]) {
    cellIdentifier = @"insert";
  } else {
    cellIdentifier = @"customProperty";
  }
  return cellIdentifier;
}

- (BOOL)isInsertRowAtIndexPath:(NSIndexPath *)indexPath {
  return indexPath.section == kPropertiesSection &&
         indexPath.row == [self tableView:self.tableView numberOfRowsInSection:indexPath.section] - 1;
}

- (BOOL)isPropertiesRowSection:(NSInteger)section {
  return section == kPropertiesSection;
}

#pragma mark - Table view delegate

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView
           editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
  if ([self isInsertRowAtIndexPath:indexPath]) {
    return UITableViewCellEditingStyleInsert;
  } else {
    return UITableViewCellEditingStyleDelete;
  }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:YES];
  if ([self isInsertRowAtIndexPath:indexPath]) {
     [self tableView:tableView commitEditingStyle:UITableViewCellEditingStyleInsert forRowAtIndexPath:indexPath];
  }
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  if ([self isPropertiesRowSection:section]) {
    return self.propertiesCount + 1;
  } else {
    return 1;
  }
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
  if ([self isPropertiesRowSection:section]) {
    return @"Properties";
  }
  return nil;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
  return [self isPropertiesRowSection:indexPath.section];
}

- (void)tableView:(UITableView *)tableView
    commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
     forRowAtIndexPath:(NSIndexPath *)indexPath {
  if (editingStyle == UITableViewCellEditingStyleDelete) {
    self.propertiesCount--;
    [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
  } else if (editingStyle == UITableViewCellEditingStyleInsert) {
    self.propertiesCount++;
    [tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
  }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  NSString *cellIdentifier = [self cellIdentifierForRowAtIndexPath:indexPath];
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
  if (cell == nil) {
    cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
  }
  return cell;
}

@end

File: TestApps/Sasquatch/Sasquatch/ViewControllers/MSCustomPropertiesViewController.swift
import UIKit

private var kPropertiesSection: Int = 0
private var kEstimatedRowHeight: CGFloat = 88.0

class MSCustomPropertiesViewController : UITableViewController, AppCenterProtocol {
  
  var propertiesCount: Int = 0
  var appCenter: AppCenterDelegate!
  
  override func viewDidLoad() {
    super.viewDidLoad()
    tableView.rowHeight = UITableViewAutomaticDimension
    tableView.estimatedRowHeight = kEstimatedRowHeight
    tableView.setEditing(true, animated: false)
    
    // Make sure the UITabBarController does not cut off the last cell.
    self.edgesForExtendedLayout = []
  }
  
  @IBAction func send() {
    let customProperties = MSCustomProperties()
    for i in 0..<propertiesCount {
      let cell = tableView.cellForRow(at: IndexPath(row: i, section: kPropertiesSection)) as? MSCustomPropertyTableViewCell
      cell?.setPropertyTo(customProperties)
    }
    appCenter.setCustomProperties(customProperties)
    
    // Clear the list.
    propertiesCount = 0
    tableView.reloadData()
    
    // Display a dialog.
    let alertController = UIAlertController(title: "The custom properties log is queued",
                                            message: nil,
                                            preferredStyle:.alert)
    alertController.addAction(UIAlertAction(title: "OK", style: .default))
    present(alertController, animated: true)
  }
  
  @IBAction func onDismissButtonPress(_ sender: Any) {
    self.dismiss(animated: true, completion: nil)
  }
  
  func cellIdentifierForRow(at indexPath: IndexPath) -> String {
    var cellIdentifier: String? = nil
    if isSendRow(at: indexPath) {
      cellIdentifier = "send"
    } else if isInsertRow(at: indexPath) {
      cellIdentifier = "insert"
    } else if isDismissRow(at:indexPath) {
      cellIdentifier = "dismiss"
    } else {
      cellIdentifier = "customProperty"
    }
    return cellIdentifier ?? ""
  }
  
  func isInsertRow(at indexPath: IndexPath) -> Bool {
    return indexPath.section == kPropertiesSection && indexPath.row == tableView(tableView, numberOfRowsInSection: indexPath.section) - 1
  }

  func isSendRow(at indexPath: IndexPath) -> Bool {
    return !isPropertiesRowSection(indexPath.section) && indexPath.row == 0
  }

  func isDismissRow(at indexPath: IndexPath) -> Bool {
    return !isPropertiesRowSection(indexPath.section) && indexPath.row == 1
  }

  func isPropertiesRowSection(_ section: Int) -> Bool {
    return section == kPropertiesSection
  }
  
  // MARK: - Table view delegate
  
  override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
    if isInsertRow(at: indexPath) {
      return .insert
    }
    else {
      return .delete
    }
  }
  
  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    tableView.deselectRow(at: indexPath, animated: true)
    if isInsertRow(at: indexPath) {
      self.tableView(tableView, commit: .insert, forRowAt: indexPath)
    }
  }
  
  // MARK: - Table view data source
  
  override func numberOfSections(in tableView: UITableView) -> Int {
    return 2
  }
  
  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    if isPropertiesRowSection(section) {
      return propertiesCount + 1
    } else {
      return 2
    }
  }
  
  override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
    if isPropertiesRowSection(section) {
      return "Properties"
    }
    return nil
  }
  
  override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
    return isPropertiesRowSection(indexPath.section)
  }
  
  override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
    if editingStyle == .delete {
      propertiesCount -= 1
      tableView.deleteRows(at: [indexPath], with: .automatic)
    } else if editingStyle == .insert {
      propertiesCount += 1
      tableView.insertRows(at: [indexPath], with: .automatic)
    }
  }
  
  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cellIdentifier = cellIdentifierForRow(at: indexPath)
    var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
    if cell == nil {
      cell = UITableViewCell(style: .default, reuseIdentifier: cellIdentifier)
    }
    return cell!
  }
}
